src/core/image_async.rs

```diff
@@ -0,0 +1,113 @@
+    /// of its arguments should be meaningful!
+    /// 2. Implementations should be conservative about which arguments they expose. Many times,
+    /// users will either go with the default arguments or just override one or two. When defining
+    /// the arguments of your image, consider that the whole purpose is to facilitate integration
+    /// testing. Only expose those that actually make sense for this case.
+    type Args;
+
+    /// A type representing the environment variables for an Image.
+    ///
+    /// There are a couple of things regarding the arguments of images:
+    ///
+    /// 1. Similar to the Default implementation of an Image, the Default instance
+    /// of its environment variables should be meaningful!
+    /// 2. Implementations should be conservative about which environment variables they expose. Many times,
+    /// users will either go with the default ones or just override one or two. When defining
+    /// the environment variables of your image, consider that the whole purpose is to facilitate integration
+    /// testing. Only expose those that actually make sense for this case.
+    type EnvVars;
+
+    /// A type representing the volumes for an Image.
+    ///
+    /// There are a couple of things regarding the arguments of images:
+    ///
+    /// 1. Similar to the Default implementation of an Image, the Default instance
+    /// of its volumes should be meaningful!
+    /// 2. Implementations should be conservative about which volumes they expose. Many times,
+    /// users will either go with the default ones or just override one or two. When defining
+    /// the volumes of your image, consider that the whole purpose is to facilitate integration
+    /// testing. Only expose those that actually make sense for this case.
+    type Volumes;
+
+    /// A type representing the entrypoint for an Image.
+    type EntryPoint: ?Sized;
+
+    /// The descriptor of the docker image.
+    ///
+    /// This should return a full-qualified descriptor.
+    /// Implementations are encouraged to include a tag that will not change (i.e. NOT latest)
+    /// in order to prevent test code from randomly breaking because the underlying docker
+    /// suddenly changed.
+    fn descriptor(&self) -> String;
+
+    /// Async function to wait until the started container is ready.
+    ///
+    /// This method is the **🍞 and butter** of the whole testcontainers library. Containers are
+    /// rarely instantly available as soon as they are started. Most of them take some time to boot
+    /// up.
+    ///
+    /// Implementations MUST return a async function signalling the current thread that the
+    /// passed-in container is ready to be interacted with. The container instance provides
+    /// access to logs of the container.
+    ///
+    /// Most implementations will very likely want to make use of this to wait for a particular
+    /// message to be emitted.
+    async fn wait_until_ready<D: DockerAsync + Sync>(
```

Comment:
This is the only function within the `ImageAsync` trait that is actually async. I think we should refactor this whole "waiting" concept in a way that works for blocking and async clients. That would allow us to reuse the same image definition for both blocking and async clients. Currently, we need to redefine every image with the `ImageAsync` trait to use it with an async client :/

One idea on how to achieve this would be to remove some flexibility in what images can do here. In particular, there seem to be primarily three things that images do to determine that they are ready:

- Wait for a message on stdout
- Wait for a message on stderr
- Wait for an arbitrary amount of time

I think we can model that be having a function that returns a `Vec<WaitFor>` where `WaitFor` is:

```rust
enum WaitFor {
	Message { message: String, stream: Stream },
	Time { ms: u64 }
}
```